src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs

```diff
@@ -52,6 +53,65 @@ public static class MemoryMarshal
+        /// <param name="source">The source slice, of type <typeparamref name="TFrom"/>.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <typeparamref name="TFrom"/> or <typeparamref name="TTo"/> contains pointers.
+        /// </exception>
+        /// <exception cref="System.OverflowException">
+        /// Thrown if the Length property of the new Span would exceed Int32.MaxValue.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<TTo> Cast<TFrom, TTo>(ReadOnlySpan<TFrom> source)
+            where TFrom : struct
+            where TTo : struct
+        {
+            if (SpanHelpers.IsReferenceOrContainsReferences<TFrom>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TFrom));
+
+            if (SpanHelpers.IsReferenceOrContainsReferences<TTo>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TTo));
+
+            int newLength = checked((int)((long)source.Length * Unsafe.SizeOf<TFrom>() / Unsafe.SizeOf<TTo>()));
+            return new Span<TTo>(Unsafe.As<Pinnable<TTo>>(source.Pinnable), source.ByteOffset, newLength);
+        }
+
+        /// <summary>
+        /// Casts a ReadOnlySpan of one primitive type <typeparamref name="TFrom"/> to another primitive type <typeparamref name="TTo"/>.
+        /// These types may not contain pointers or references. This is checked at runtime in order to preserve type safety.
+        /// </summary>
+        /// <remarks>
+        /// Supported only for platforms that support misaligned memory access.
+        /// </remarks>
+        /// <param name="source">The source slice, of type <typeparamref name="TFrom"/>.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <typeparamref name="TFrom"/> or <typeparamref name="TTo"/> contains pointers.
+        /// </exception>
+        /// <exception cref="System.OverflowException">
+        /// Thrown if the Length property of the new Span would exceed Int32.MaxValue.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<TTo> Cast<TFrom, TTo>(ReadOnlySpan<TFrom> source)
+            where TFrom : struct
+            where TTo : struct
+        {
+            if (SpanHelpers.IsReferenceOrContainsReferences<TFrom>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TFrom));
+
+            if (SpanHelpers.IsReferenceOrContainsReferences<TTo>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TTo));
+
+            int newLength = checked((int)((long)source.Length * Unsafe.SizeOf<TFrom>() / Unsafe.SizeOf<TTo>()));
+            return new ReadOnlySpan<TTo>(Unsafe.As<Pinnable<TTo>>(source.Pinnable), source.ByteOffset, newLength);
+        }
+
 #else
```

Comment:
Why do we have both the portable and fast versions in the same file?  Can we split this into separate files like we do elsewhere?

Reply:
The purpose of the `Shared` file is to make it so that merging the coreclr PR won't cause CoreFX to break. To do that, we need to mostly leave the existing `MemoryMarshal` file intact, which means we can't include it in coreclr if we also have a `MemoryMarshal.Fast` that contains some of the signatures that are currently living in `MemoryMarshal`.

Reply:
There are two cases:
- You are adding new APIs (including adding overloads of existing APIs with new signatures). In this case, you can just add it to MemoryMarshal.cs
- You add modifying existing APIs. In this case, you will need to augment the mirror PR in CoreFX to get it through. `MemoryMarshal.Shared.cs` file is not going to help you with that.

What I am missing?

Reply:
>You add modifying existing APIs. In this case, you will need to augment the mirror PR in CoreFX to get it through. MemoryMarshal.Shared.cs file is not going to help you with that.

The way I have it currently set up, I'm not actually modifying any existing APIs, I'm only adding new API, so the mirror PR won't be broken.

> You are adding new APIs (including adding overloads of existing APIs with new signatures). In this case, you can just add it to MemoryMarshal.cs

If I add the new API to MemoryMarshal, then coreclr needs to include MemoryMarshal, which means I'll need to remove the Fast-Specific section from MemoryMarshal to avoid conflicts with MemoryMarshal.Fast.cs, which means that the mirror PR to corefx will be broken for the non-portable builds unless I modify the mirror PR to include the MemoryMarshal.Fast file. 

I was trying to avoid having to make fixes to the mirror PR, but if you're strongly against the `Shared` suffix, then I can. I'll just fix the mirror PR to include the new MM.Fast and MM.Portable files to get the build green then keep my feature CoreFX PR the same minus the new file inclusions.

Reply:
>  you're strongly against the Shared suffix

Yes, I am strongly against adding temporary files with .Shared suffix. It just means more round trips between repos to clean things up.

> mirror PR to corefx will be broken for the non-portable builds 

We should expect that you need to patch up the mirror PR for changes like these.

Reply:
👍 Shared removed.